_posts/2023-02-22-user-defined-operators.md

```diff
@@ -0,0 +1,484 @@
+1. More type-safety when performing operations on values of the same underlying value type that express
+    conceptually different things.
+    For example quantities of different tokens could both be stored as `uint` but adding them without
+    an explicit conversion is a mistake that can only be prevented if the compiler knows that they
+    are values of different kinds.
+
+1. The ability to define new types using value types as the underlying representation.
+    The best example of this are fixed-point types.
+
+    Fixed-point types were originally going to be an integral part of the language, and the
+    partially implemented [`ufixed` and `fixed`](https://docs.soliditylang.org/en/latest/types.html#fixed-point-numbers)
+    types are the vestiges of that old design.
+    While trying to agree on the final design for the feature, however, we realized that no implementation
+    is going to be a trade-off good enough for most situations and cannot become a suitable default choice.
+    We decided to leave the choice up to library authors and instead focus on providing primitives
+    necessary to build their own fixed-point types that feel like a native type.
+
+#### Will User-Defined Operators Make Auditing Harder?
+In the feedback we received so far one of the most common concerns is the effect of the feature on contract
+auditing and potential to mask malicious code.
+I want to address these concerns here.
+
+First of all, any new syntax increases the complexity of the language and has some potential to make code more obscure
+and harder to understand when abused.
+It cannot be completely avoided and is always a trade-off between all the good and bad ways it can be used.
+We think that improving the ergonomics of UDVTs is important enough that it outweights any of the downsides we considered.
+
+It is important to realize the actual scope of the feature.
+You may be imagining an elaborate system from languages like C++, with complicated rules and many corner cases.
+As implemented now the feature is deliberately very limited.
+We tried to both limit surprises and avoid prematurely adding support for cases that we have not carefully thought out.
+
+In the current implementation it is not possible to:
+
+1. Have operators on types other than UDVTs.
+    - User-defined operators on [reference types](https://docs.soliditylang.org/en/latest/types.html#reference-types)
+      like structs or arrays are not allowed.
+    - It is not possible to define them on built-in value types, function types or enums.
+    - You cannot define them to unnamed types using the type wildcard (`*`).
+1. Have operators that are not `pure`.
+    Operators cannot modify storage.
+    They can at most make pure external calls.
+
+    The worst that is technically possible with tricks is for them to modify memory (only through inline assembly)
+    or make staticcalls (through interfaces that do not match actual code deployed on chain).
```

Comment:
```suggestion
In the current implementation, it is not possible to:

1. Have operators on types other than UDVTs.
    - User-defined operators on [reference types](https://docs.soliditylang.org/en/latest/types.html#reference-types)
      like structs or arrays are not allowed.
    - It is not possible to define them on built-in value types, function types or enums.
    - You cannot define them to unnamed types using the type wildcard (`*`).
1. Have operators that are not `pure`.
    Operators cannot modify storage.
    They can, at most, make pure external calls.

    The worst that is technically possible with tricks is for them to modify memory (only through inline assembly)
    or make staticcalls (through interfaces that do not match actual code deployed on-chain).
```

Reply:
Is `on-chain` actually correct here? I mean it's deployed on the chain, not using it as an adjective.

In any case, I reworded the sentence to remove the word :)